Enhance Mistral model support with reasoning capabilities

colesmcintosh opened the pull request:
## Title

Enhance Mistral model support with reasoning capabilities

## Relevant issues

<!-- Add relevant issue numbers if any -->

## Pre-Submission checklist

**Please complete all items before asking a LiteLLM maintainer to review your PR**

- [x] I have Added testing in the [`tests/litellm/`](https://github.com/BerriAI/litellm/tree/main/tests/litellm) directory, **Adding at least 1 test is a hard requirement** - [see details](https://docs.litellm.ai/docs/extras/contributing_code)
- [x] I have added a screenshot of my new test passing locally 

<img width="1257" alt="Screenshot 2025-06-11 at 5 19 35 PM" src="https://github.com/user-attachments/assets/e4fc2e80-3894-4ddd-a623-fd5e43c13486" />

- [x] My PR passes all unit tests on [`make test-unit`](https://docs.litellm.ai/docs/extras/contributing_code)
- [x] My PR's scope is as isolated as possible, it only solves 1 specific problem

## Type

🆕 New Feature

## Changes

This PR adds comprehensive support for Mistral's reasoning capabilities, specifically for magistral models that support advanced reasoning features.

### Key Features Added:

1. **Reasoning Parameter Support**: Added support for `reasoning_effort` and `thinking` parameters for magistral models
2. **System Prompt Injection**: Automatically injects reasoning system prompts that instruct the model to use `<think>` tags for step-by-step reasoning
3. **Model-Specific Handling**: Only applies reasoning features to magistral models (case-insensitive detection)
4. **Seamless Integration**: Works with existing LiteLLM parameter mapping and transformation pipeline

### Files Modified:

- **`litellm/llms/mistral/mistral_chat_transformation.py`**: Enhanced MistralConfig class with reasoning support
  - Added `get_supported_openai_params()` method to expose reasoning parameters for magistral models
  - Added `map_openai_params()` enhancement to handle reasoning parameters
  - Added `_add_reasoning_system_prompt_if_needed()` method for prompt injection
  - Added `_get_mistral_reasoning_system_prompt()` method for generating reasoning prompts

- **`model_prices_and_context_window.json`**: Updated model definitions to reflect reasoning capabilities

- **`tests/test_litellm/llms/mistral/test_mistral_chat_transformation.py`**: Comprehensive test suite (249 lines)
  - Tests parameter mapping for magistral vs non-magistral models
  - Tests system prompt injection with and without existing system messages
  - Tests case-insensitive model detection
  - Tests end-to-end reasoning workflow
  - Tests that non-magistral models properly ignore reasoning parameters

### Technical Implementation:

- **Parameter Mapping**: `reasoning_effort` and `thinking` parameters are mapped to an internal `_add_reasoning_prompt` flag
- **Prompt Template**: Uses `<think>` and `</think>` tags as specified by Mistral's reasoning implementation
- **System Message Handling**: Intelligently adds or modifies system messages to include reasoning instructions
- **Backward Compatibility**: Non-magistral models continue to work unchanged

### Testing:

All tests pass (13/13) with comprehensive coverage of:
- Parameter validation and mapping
- System prompt injection logic
- Model type detection
- Integration workflows
- Edge cases and error handling

This enhancement enables users to leverage Mistral's advanced reasoning capabilities through LiteLLM's unified interface while maintaining full backward compatibility.

colesmcintosh committed:
[Feat] Enhance Mistral model support with reasoning capabilities

* Added support for reasoning parameters in magistral models, including "reasoning_effort" and "thinking".
* Updated the MistralConfig class to handle reasoning system prompts.
* Implemented tests to verify reasoning functionality and ensure correct parameter mapping for magistral models.
* Enhanced the model prices JSON to reflect new reasoning capabilities.

colesmcintosh committed:
[Refactor] Improve type safety in Mistral chat transformation

* Added type casting for messages to enhance type safety.
* Updated the handling of system messages to ensure correct content formatting.
* Cleaned up code for better readability and maintainability.

colesmcintosh committed:
[Feat] Add Mistral AI reasoning capabilities docs

ishaan-jaff commented on docs/my-website/docs/providers/mistral.md:
```diff
@@ -144,20 +144,22 @@ All models listed here https://docs.mistral.ai/platform/endpoints are supported.
-| Codestral Mamba    | `completion(model="mistral/codestral-mamba-latest"", messages)`     |
+| Model Name     | Function Call                                                | Reasoning Support |
+|----------------|--------------------------------------------------------------|-------------------|
+| Mistral Small  | `completion(model="mistral/mistral-small-latest", messages)` | ❌ |
```
can we limit the emojis here, feels overwhelming

ishaan-jaff commented on litellm/llms/mistral/mistral_chat_transformation.py:
```diff
@@ -205,6 +231,43 @@ def _transform_messages(
+            for i, msg in enumerate(messages):
+                if msg.get("role") == "system":
+                    existing_content = msg.get("content", "")
+                    reasoning_prompt = self._get_mistral_reasoning_system_prompt()
```
can you handle when the content is a list ?

ishaan-jaff reviewed (changes requested):
reviewed

colesmcintosh committed:
Merge branch 'BerriAI:main' into mistral-reasoning

colesmcintosh committed:
Update Mistral documentation and enhance reasoning prompt handling

- Revised the reasoning support indicators in the Mistral model documentation for clarity.
- Improved the `_add_reasoning_system_prompt_if_needed` method to handle both string and list content types for system messages, ensuring the reasoning prompt is correctly prepended.
- Added a new test case to verify the functionality of adding the reasoning system prompt when the existing content is a list.

colesmcintosh committed:
Refactor Mistral chat transformation to handle list content

- Updated the `_add_reasoning_system_prompt_if_needed` method to convert list content to strings before prepending the reasoning prompt.
- Adjusted tests to verify that system messages with list content are correctly transformed into strings, ensuring original content is preserved.

colesmcintosh committed:
Enhance Mistral chat transformation to preserve content types

- Updated the `_add_reasoning_system_prompt_if_needed` method to maintain the original format of list content when prepending the reasoning prompt.
- Adjusted tests to verify that both string and list content types are correctly handled, ensuring the reasoning prompt is added without altering the content structure.

colesmcintosh committed:
fix linting error

ishaan-jaff commented on docs/my-website/docs/providers/mistral.md:
````diff
@@ -203,6 +205,110 @@ assert isinstance(
+
+The `reasoning_effort` parameter controls how much effort the model puts into reasoning. When used with magistral models.
+
+```python
````
nit: can you add showLineNumbers for all code snippets on docs. Trying to get more docs like this as it looks polished

ishaan-jaff reviewed (approved):
LGTM !

krrishdholakia commented on litellm/llms/mistral/mistral_chat_transformation.py:
```diff
@@ -87,6 +87,12 @@ def get_supported_openai_params(self, model: str) -> List[str]:
+        
+        # Add reasoning support for magistral models
+        if "magistral" in model.lower():
+            supported_params.extend(["thinking", "reasoning_effort"])
```
'thinking' is said to be supported - but i don't see that param mapped @colesmcintosh
